Awaitables

fleming79 opened the pull request:
This PR is a subset of #160

This PR enables asynchronous code execution as follows:

- Python methods are coroutines (`async def`)
- Each request sent to the frontend is assigned a `Pending` (just like an asyncio.Future). When the frontend replies, the pending is set done and the result is returned.
- A JSON representation is returned if the result can be converted to JSON.
- A Connection object is provided if the object returned in the frontend is a disposable (or when the transform is specified as a connection).
- The connection can be used to:
    -  access the methods and properties 
    - close the object
    - use the object as a reference in the frontend using 'toObject'
    - conveted to a lumino widget using 'toLuminoWidget'

### Demo running all cells at once

Awaiting in each cell ensures the result is available for the next cell. The gif below is 0.5x playback speed.

![ezgif-405a2e1b8e1514bb](https://github.com/user-attachments/assets/556213ee-ad6e-4dca-bf1e-ff22bf10fa0c)

### Menus

This gif demonstrates the usage of menus and restoration when the page is refreshed.

![ezgif-4b4a42d822193ae5](https://github.com/user-attachments/assets/9131d96d-75df-4bfc-a820-4ce1764b297d)


Async execution is enabled in notebooks with [async-kernel](https://github.com/fleming79/async-kernel).  


### Subshells


**Requires async-kernel**

Leverages on the concept of *kernel subshells* to provide isolated `user_ns` and common `global_user_ns`.


![ezgif-36277de683b32f49](https://github.com/user-attachments/assets/6ee1944b-f981-4344-b425-cd2da6c6dd30)

resolves #11 

Replaces #160

fleming79 commented:
@jtpio I've made an attempt at splitting up the the other PR into smaller chunks. Unfortunately, it still remains rather large.

Feel free to make changes as you see fit, particularly with the js code.

Notes:
- I switched to using uv for building to make it is easier to setup a dev environment on Windows. 
- I don't think the async version will be compatible with Jupyterlite.

Commit:
V2 release :
- All methods are async
- widget restoration on page refresh
- Preliminary support for multiple pages
- Added python classes
  - Ipylab
  - Singular
  - HasApp
  - Fixed
  - Typing instances
  - Transform
  - Connection

Commit:
Bump async_kernel version which has dropped the AsyncEvent class and adopted aiologic for synchronisation primitives (Event & Lock).

Commit:
Bump async-kernel min version to 0.7 and update lock file.

fleming79 committed:
Breaking changes for compatibility with async-kernel v9.

'Fixed' and its type hints have been transferred to async-kernel.common.
Future has been renamed to Pending and Caller.get_instance has been renamed to Caller.get.

fleming79 committed:
Bump async-kernel min version to 0.9rc1.

fleming79 committed:
Added  "Programming Language :: Python :: 3.14",

fleming79 committed:
import_item was also moved to async_kernel.

Commit:
Bump async-kernel version.

Commit:
Add described_by to CommandRegistry.

fleming79 commented:
`described_by` was added in [this commit](https://github.com/jtpio/ipylab/pull/161/commits/70bb82e66c19c8bf7e187cff755c6c3c499e27c9). It provides a similar interface to the first part of #162 (printing args only). 

<img width="1878" height="923" alt="image" src="https://github.com/user-attachments/assets/9bc5b76e-9cfc-4665-9d8d-ca2f2518d78a" />

This illustrates how awaitables can be used to add in more functionality.

bollwyvl commented:
Yeh, the goal over on #162 was to add some more optional features without breaking the public APIs (even though _that_ took a good deal of mangling). The `execute(..., handler)` is likely even more interesting than `describe`. 

Until there is a proper upstream `ipywidgets` async (iterable) story, I think having additive, alternate APIs here would be nice, for sure.

Commit:
Remove patched modules - not required for this branch to work.

fleming79 commented:
> Yeh, the goal over on #162 was to add some more optional features without breaking the public APIs (even though _that_ took a good deal of mangling). The `execute(..., handler)` is likely even more interesting than `describe`.
> 
> Until there is a proper upstream `ipywidgets` async (iterable) story, I think having additive, alternate APIs here would be nice, for sure.

Your comment *"Until there is a proper upstream ipywidgets async"* made me realize that the patched libraries aren't required for this PR; so I've removed them. At the moment it still requires the kernel to be [async-kernel](https://pypi.org/project/async-kernel/), which was developed specifically to [make ipylab work nicely in notebooks](https://github.com/ipython/ipykernel/pull/1384#issuecomment-3205367767).


In this PR:
-  [`CommandRegistry.execute`](https://github.com/jtpio/ipylab/pull/161/files#diff-59f42ab965c5f5c812ec649e466bcc96585f08e7af00895e246c370237af2a8dR279) returns the result of execution in the frontend (or a `Connection` to an object).
- The 'described_by' feature (without validation) is configurable on the [CommandConnection](https://github.com/jtpio/ipylab/pull/161/files#diff-59f42ab965c5f5c812ec649e466bcc96585f08e7af00895e246c370237af2a8dR81) which gets returned when a command is created. 
- Icons can be included in [`CommandRegistry.add_command`](https://github.com/jtpio/ipylab/pull/161/files#diff-59f42ab965c5f5c812ec649e466bcc96585f08e7af00895e246c370237af2a8dR227)
 - Multiple command registries are supported; a new registry can be created with `CommandRegistry(name="my new registry")`.

Commit:
Bump async-kernel min version to 0.9.1.

Commit:
Strip out unrequired features
- log.py
- incorrectly included hookimpl

Commit:
Remove namespace_id from JupyterFrontEnd.

Commit:
Remove namespace_id from CommandRegistry and simplify.

Commit:
Provide default icon values.

Commit:
- Drop concept of page_id (could be added later).
- Add subshell support with greater dependence on async kernel.

fleming79 committed:
Fix icons - remove duplicate definition and ensure there is only one LabIcon for the life of the Icon.

fleming79 committed:
Move launcher closing code to the frontend.

fleming79 committed:
Ipylab register on_msg before calling __init__.

fleming79 committed:
Better CommandPalette validation.

fleming79 committed:
Ensure Connection id looks valid.

fleming79 committed:
Bump async-kernel min version to 0.10.1.

Commit:
Improve type hinting.

fleming79 committed:
Make transform_payload non-async and don't wait for connection to be ready.

fleming79 committed:
Remove advanced and function features from transform.

Commit:
Remove widget restoration capability (still avaliable in the full version).

Commit:
validate_command_id should wait for command to be ready.

Commit:
Compatibility with Jupyterlite using jupyterlite_async_kernel.

Commit:
Modify readythedocs.yml

Commit:
Fix tests.

Commit:
Make jupyterlite path conditional in readme.

Commit:
Make it more obvious that a PR is viewable.

fleming79 commented:
I've been working on a Jupyterlite plugin for 'async-kernel' which now mostly working. For the moment the code is being hosted here: https://github.com/fleming79/echo-kernel.

The interface is simple in design and should be pretty robust. It uses a web worker to run pyodide which loads and runs the kernel. Messages are transferred using the builtin post functions and kernel messages are passed via a single callback function on each side, the messages are serialized to json for` javascript <--> python` conversion.

The kernel is started with a script which is configurable, by default script searches for wheels locally, and installs them prior to starting the kernel. This helps to make it easier to bundle wheels in jupyterlite.

There are a few associated PRs about interrupts, which is the main missing feature.

- https://github.com/jupyterlab/jupyterlab/pull/18307
- https://github.com/jupyterlite/jupyterlite/pull/1834

I've adapted ipylab readthedocs to use the new kernel. You can access the most recent docs [here](https://ipylab--161.org.readthedocs.build/en/161/lite/lab/).

Commit:
Remove the script from the readme, which doesn't render on github.

RDT preview is a an alternative.

https://github.com/jupyterlite/jupyterlite/blob/main/.github/workflows/rtd-preview.yml

Commit:
Bump async-kernel version range to l>=0.12,<0.13

Commit:
Remove unused "initializing" case.

Commit:
Bump async-kernel min version to 12.3.

Commit:
Improve description in subshells.ipynb.

Commit:
Bump async-kernel min version to 0.14.

Commit:
Bugfix __init_subclass__ should always call super().__init_subclass__(**kwargs) : see https://discuss.python.org/t/type-object-cls-has-no-attribute-parameters/104276/4.

Commit:
Update uv.lock.

Commit:
Bump async-kernel min version to 0.14.

Commit:
Upgrade uv lock file.

Commit:
Bump async-kernel version range to >0.16.0 < 0.17.0.
- async-kernel no uses a standard dict instead of a last executed dict for better performance.
- CatchResult is added in JupyterFrontEnd._evaluated to provide the necessary functionality.
- "payload" no longer has any meaning in JupyterFrontEnd.evaluate.

Commit:
Bump async-kernel version range to async-kernel>=0.17,<0.18.

Commit:
Bump async-kernel range to async-kernel>=0.18,<0.19

Commit:
Update uv.lock.

Commit:
async-kernel compatibility version 0.19.0

Commit:
Type checking updates.

Commit:
Upgrade uv.lock.

Commit:
Rename mic to connection.

Commit:
Run pre-commit.

Commit:
Fix subshells.ipynb

Commit:
Bump async-kernel min version to 0.20.0 and use pack_json_str from async_kernel.compat.json.

Commit:
Remove @jupyter/ydoc dependency.

Commit:
Update uv.lock.

Commit:
Switch from @jupyterlab/builder to @jupyter/builder

Commit:
Update js deps

Commit:
Add jupyterlite as docs dep and add specific traitlets dep.

Commit:
Replace jupyterlab with jupyter_builder for build-system.

Commit:
Update tsconfig.json

Commit:
Update uv.lock.

Commit:
Pre-commit autoupdate

Commit:
Allow sending buffers to the frontend in operation.

Commit:
Drop wait_ready in notebooks.

Commit:
Update docstring and use aiologic.meta.iscoroutinelike instead of inspect.iscoroutine.

Commit:
Swap aiologic.BinarySemaphore for aiologic.Lock in CommandRegistry._lock.

Commit:
Bump async-kernel version range to async-kernel>=0.21,<0.22.

Commit:
Upgrade uv.lock.

Commit:
Upgrade pytest.

Commit:
Pre-commit autoupdate.

Commit:
Add uv min version.

Commit:
Add notes about Jupyterlite in README.md.

Commit:
Update the jupyterlite-async-kernel version.

Commit:
Update build.yml for uv version specified in pyproject file.
